### PR TITLE
[DependencyInjection] Use lazy-loading ghost object proxies out of the box

### LINKED ISSUE #35345: ProxyManagerBridge should possibly use Lazy Loading Ghost Object Proxies instead of Lazy Loading Value Holder Proxies
I have created new issue to make it less doctrine specific.

Here is demonstration of different lazy proxies (modified [proxy-manager examples]( https://github.com/Ocramius/ProxyManager/tree/master/examples))

```php

use ProxyManager\Factory\LazyLoadingGhostFactory;
use ProxyManager\Factory\LazyLoadingValueHolderFactory;

class Foo
{
    public function compare(self $foo)
    {
        var_dump($foo === $this);
    }
}


$proxy = (new LazyLoadingValueHolderFactory())->createProxy(Foo::class, function (&$wrappedObject) {
    $wrappedObject = new Foo();
    return true;
});

$proxy->compare($proxy); // bool(false)

$proxy = (new LazyLoadingGhostFactory())->createProxy(Foo::class, fn() => true);
$proxy->compare($proxy); // bool(true)
```

First approach is used by Symfony. As you can see, it breaks identity and is more complicated to use. I suggest to use second method, which doesn't break identities. Would solve https://github.com/symfony/symfony/issues/35216 and issues related to it.

cc @stof

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #35345
| License       | MIT
| Doc PR        | -

This PR builds on #46751. It also replaces #46458.

Instead of using ProxyManager to make lazy services actually lazy, using `LazyGhostObjectTrait` from #46751 allows doing so *out of the box* - aka without the need to install any optional dependencies.

When a virtual proxy is required (typically when using [the `proxy` tag](https://github.com/symfony/symfony/pull/27697)), ProxyManager is still required (and the dep remains optional.)

But for most services, using `LazyGhostObjectTrait` just works \o/

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/LazyProxy/Instantiator/LazyServiceInstantiator.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    public function instantiateProxy(ContainerInterface $container, Definition $definition, string $id, callable $realInstantiator): object
+    {
+        $dumper = new LazyServiceDumper();
+
+        if ($dumper->useProxyManager($definition)) {
+            return (new RuntimeInstantiator())->instantiateProxy($container, $definition, $id, $realInstantiator);
+        }
+
+        if (!class_exists($proxyClass = $dumper->getProxyClass($definition), false)) {
+            eval(sprintf('class %s extends %s implements %s { use %s; }', $proxyClass, $definition->getClass(), LazyGhostObjectInterface::class, LazyGhostObjectTrait::class));
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/LazyProxy/Instantiator/LazyServiceInstantiator.php:38]: what about environments where `eval` is disabled ?

File: src/Symfony/Component/DependencyInjection/LazyProxy/PhpDumper/LazyServiceDumper.php [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            return false;
+        }
+
+        if (!($class = $definition->getClass()) || !(class_exists($class) || interface_exists($class, false))) {
+            return false;
+        }
+
+        $class = new \ReflectionClass($class);
+
+        if ($asGhostObject = !$class->isAbstract() && !$class->isInterface() && ('stdClass' === $class->name || !$class->isInternal())) {
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/LazyProxy/PhpDumper/LazyServiceDumper.php]: what about final classes ? shouldn't we have a clean error handling ?

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/LazyProxy/PhpDumper/LazyServiceDumper.php]: I added a check to throw an exception for final classes.

File: src/Symfony/Component/DependencyInjection/LazyProxy/Instantiator/LazyServiceInstantiator.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/LazyProxy/Instantiator/LazyServiceInstantiator.php:38]: AFAIK `eval` cannot be disabled unless a fancy extension does so. Suhosin used to be such an extension but it's not available on PHP 8.1 IIUC. I wouldn't care much. I'd wait for someone to be affected before caring at all.

File: src/Symfony/Component/DependencyInjection/LazyProxy/PhpDumper/LazyServiceDumper.php [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            return false;
+        }
+
+        $class = new \ReflectionClass($class);
+
+        if ($class->isFinal()) {
+            throw new InvalidArgumentException(sprintf('Cannot make service of class "%s" lazy because the class is final.', $definition->getClass()));
+        }
+
+        if ($asGhostObject = !$class->isAbstract() && !$class->isInterface() && ('stdClass' === $class->name || !$class->isInternal())) {
```
Comment (Tobion) [inline src/Symfony/Component/DependencyInjection/LazyProxy/PhpDumper/LazyServiceDumper.php]: I would use `\stdClass::class` to make it semantic and detectable.

File: src/Symfony/Component/VarExporter/Tests/Fixtures/LazyGhostObject/MagicClass.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\VarExporter\Tests\Fixtures\LazyGhostObject;
+
+class MagicClass
+{
+    public static $destructCounter = 0;
+    public $cloneCounter = 0;
```
Comment (Tobion) [inline src/Symfony/Component/VarExporter/Tests/Fixtures/LazyGhostObject/MagicClass.php:17]: missing property types?

File: src/Symfony/Component/DependencyInjection/LazyProxy/PhpDumper/LazyServiceDumper.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/LazyProxy/PhpDumper/LazyServiceDumper.php]: alright, updated

**Comment by nicolas-grekas** [comment]
PR rebased and ready \o/

**Review by chalasr — APPROVED** [review]

**Pull request merged**